sdk/core/azure-core/src/main/java/com/azure/core/implementation/ResponseConstructorsCache.java

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.implementation;
+
+import com.azure.core.http.HttpHeaders;
+import com.azure.core.http.HttpRequest;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.rest.Response;
+import com.azure.core.implementation.serializer.HttpResponseDecoder;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.Exceptions;
+import reactor.core.publisher.Mono;
+
+import java.lang.invoke.LambdaMetafactory;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.Constructor;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * A concurrent cache of {@link Response} constructors.
+ */
+final class ResponseConstructorsCache {
+    private final ClientLogger logger = new ClientLogger(ResponseConstructorsCache.class);
+    private final Map<Class<?>, Optional<ResponseConstructor>> cache = new ConcurrentHashMap<>();
+
+    /**
+     * Identify the most specific constructor for the given response class.
+     *
+     * The most specific constructor is looked up following order:
+     * 1. (httpRequest, statusCode, headers, body, decodedHeaders)
+     * 2. (httpRequest, statusCode, headers, body)
+     * 3. (httpRequest, statusCode, headers)
+     *
+     * @param responseClass the response class
+     * @return optional with located constructor, empty optional if there is no match
+     */
+    Optional<ResponseConstructor> get(Class<? extends Response<?>> responseClass) {
+        this.cache.computeIfAbsent(responseClass, responseCls -> Arrays.stream(responseCls.getDeclaredConstructors())
+            .filter(constructor -> {
+                int paramCount = constructor.getParameterCount();
+                return paramCount >= 3 && paramCount <= 5;
+            })
+            .sorted(Comparator.comparingInt(Constructor::getParameterCount))
+            .findFirst()
+            .flatMap(constructor -> {
+                try {
+                    MethodHandles.Lookup lookup = MethodHandles.lookup();
```

Comment:
We discussed moving this out and reusing it - is that possible?

Reply:
this is something I'm still exploring, the scope of `MethodHandles.lookup`.

Reply:
yes possible, switched to singleton Lookup.